### ISSUE TITLE
documentation of `tp_dealloc` default is unclear

### ISSUE THREAD

**Issue by davidhewitt** [python/cpython#105018]
# Documentation

The [documentation for `tp_dealloc`](https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_dealloc) suggests that setting the `tp_dealloc` slot is required:

> This function must be defined unless the type guarantees that its instances will never be deallocated (as is the case for the singletons None and Ellipsis).

Actually this seems incorrect. `PyType_FromSpec` will set all types to use a function `subtype_dealloc` if they do not define a `tp_dealloc`.

https://github.com/python/cpython/blob/1ce801b81ce63867ce382f6e9f56873a844c2bc6/Objects/typeobject.c#L2971

This seems to even be known, as implied by https://github.com/python/cpython/issues/71166#issuecomment-1093714420

It looks like it has been this way for the whole of Python 3's history.

I'm considering relying on this default for PyO3 to avoid needing to support our own `tp_dealloc` implementation. Is that a sound thing to do?

cc @encukou @serhiy-storchaka @cstimer as you were participants of #71166

**Comment by encukou** [comment]
Yes, it should be safe if you don't need custom deallocation. (Note that you can't rely on `tp_clear`, it's not always called.)
Unfortunately the slot docs were written with static types in mind. Improvements welcome.
